File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/42-TrappingRainWater/42-TrappingRainWater.java
// Last updated: 8/4/2026, 7:24:54 PM
class Solution {
    public int trap(int[] height) {
        int l_max = 0;
        int r_max = 0;
        int curr = 0;
        int ans = 0;
        Stack<Integer>st = new Stack<>();
        for(int i = 0 ; i < height.length ; i++){
            while(!st.empty() && height[i] > height[st.peek()]){
                r_max = i;
                curr = st.pop();
                if(st.empty()) break;
                l_max = st.peek();
                int width = r_max - l_max - 1;

                ans += (Math.min(height[l_max],height[r_max]) - height[curr]) * width;
            }
            st.push(i);
        }
        return ans;
    }
}

File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/42-TrappingRainWater/42-TrappingRainWater.cpp
// Last updated: 8/4/2026, 7:24:46 PM
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int leftMax = 0 , rightMax = 0 , l = 0 , r = n-1;
        int trappedRain = 0;

        while(l < r){
            if(height[l] <= height[r]){
                if(leftMax > height[l]){
                    trappedRain += leftMax - height[l];
                }
                else{
                    leftMax = height[l];
                }
                l++;
            }
            else{
                if(rightMax > height[r]){
                    trappedRain += rightMax - height[r];
                }
                else{
                    rightMax = height[r];
                }
                r--;
            }
        }
        return trappedRain;
    }
};